sdk/Pulumi/Deployment/Deployment_Prepare.cs

```diff
@@ -146,6 +136,183 @@ void LogExcessive(string message)
             }
         }
 
+        static async Task<T> Resolve<T>(Input<T>? input, T whenUnknown) 
+        {
+            return input == null
+                ? whenUnknown
+                : await input.ToOutput().GetValueAsync(whenUnknown).ConfigureAwait(false);
+        }
+
+        private static async Task<List<Pulumirpc.Alias>> PrepareAliases(
+            Resource resource, 
+            ResourceOptions options,
+            bool resourceMonitorSupportsAliasSpec)
+        {
+            var aliases = new List<Pulumirpc.Alias>();
+            if (resourceMonitorSupportsAliasSpec)
+            {
+                // map each alias in the options to its corresponding Alias proto definition.
+                foreach (var alias in options.Aliases)
+                {
+                    var resolvedAlias = await alias.ToOutput().GetValueAsync(whenUnknown: default!).ConfigureAwait(false);
+                    if (resolvedAlias == null)
+                    {
+                        // alias contains unknowns, skip it.
+                        continue;
+                    }
+
+                    var aliasName = await Resolve(resolvedAlias.Name, whenUnknown: "");
+                    var aliasType = await Resolve(resolvedAlias.Type, whenUnknown: "");
+                    var aliasStack = await Resolve(resolvedAlias.Stack, Instance.StackName);
+                    var aliasProject = await Resolve(resolvedAlias.Project, Instance.ProjectName);
+
+                    if (string.IsNullOrEmpty(aliasName) || string.IsNullOrEmpty(aliasType))
```

Comment:
This doesn't seem right. Quite often you'll give an alias with just a name and so type will be ""

Reply:
In `CollapseAliasToUrn` we have
```csharp
var name = a.Name ?? defaultName;
var type = a.Type ?? defaultType;

// later... 

if (name == null)
    throw new ArgumentNullException("No valid 'Name' passed in for alias.");
if (type == null)
    throw new ArgumentNullException("No valid 'type' passed in for alias.");
```
Which suggested that both `name` _and_ `type` must be non-null so I wasn't sure. I will change it to:
```csharp
var aliasSpec = new Pulumirpc.Alias.Types.Spec
{
    Name = await Resolve(resolvedAlias.Name, resource.GetResourceName()),
    Type = await Resolve(resolvedAlias.Type, resource.GetResourceType()),
    Stack = await Resolve(resolvedAlias.Stack, Instance.StackName),
    Project = await Resolve(resolvedAlias.Project, Instance.ProjectName),
};
```

Reply:
I'm pretty sure the engine is happy to take an Alias object with type or name as "".